src/python/pants/goal/anonymous_telemetry.py

```diff
@@ -0,0 +1,185 @@
+            help=(
+                f"An anonymized ID representing this repo.  For private repos, you likely want the "
+                f"ID to not be derived from, or algorithmically convertible to, anything "
+                f"identifying the repo.  For public repos the ID may be visible in that repo's "
+                f"config file, so anonymity of the repo is not guaranteed (although user anonymity "
+                f"is always guaranteed). {_telemetry_docs_referral}."
+            ),
+        )
+
+    @property
+    def enabled(self) -> bool:
+        return cast(bool, self.options.enabled)
+
+    @property
+    def repo_id(self) -> str | None:
+        return None if self.options.repo_id is None else cast(str, self.options.repo_id)
+
+
+class AnonymousTelemetryCallback(WorkunitsCallback):
+    def __init__(self, anonymous_telemetry: AnonymousTelemetry) -> None:
+        super().__init__()
+        self._anonymous_telemetry = anonymous_telemetry
+
+    @staticmethod
+    def validate_repo_id(repo_id: str) -> bool:
+        is_valid = re.match(r"^[a-zA-Z0-9-_]{30,60}$", repo_id) is not None
+        if not is_valid:
+            logger.error(
+                "The repo_id must be between 30 and 60 characters long, and consist of only "
+                "alphanumeric characters, dashes and underscores."
+            )
+        return is_valid
+
+    def __call__(
+        self,
+        *,
+        started_workunits: tuple[Workunit, ...],
+        completed_workunits: tuple[Workunit, ...],
+        finished: bool,
+        context: StreamingWorkunitContext,
+    ) -> None:
+        if not finished:
+            return
+
+        if self._anonymous_telemetry.options.is_default("enabled"):
+            logger.warning(
+                f"Please either set `enabled = true` in the [anonymous-telemetry] section of "
+                f"pants.toml to enable sending anonymous stats to the Pants project to aid "
+                f"development, or set `enabled = false` to disable it. No telemetry sent "
+                f"for this run. An explicit setting will get rid of this message. "
+                f"{_telemetry_docs_referral}."
+            )
+
+        if self._anonymous_telemetry.enabled:
+            repo_id = self._anonymous_telemetry.repo_id
+            if repo_id is None:
+                logger.error(
+                    f'Please set `repo_id = "<uuid>"` in the [anonymous-telemetry] section '
+                    f"of pants.toml, where `<uuid>` is some fixed random identifier, such as "
+                    f"one generated by uuidgen. No telemetry sent for this run. "
+                    f"{_telemetry_docs_referral}."
+                )
+            elif self.validate_repo_id(repo_id):
+                # Assemble and send the telemetry.
+                # Note that this method is called with finished=True only after the
+                # StreamingWorkunitHandler.session() ends, i.e., after end_run() has been called,
```

Comment:
Stale, `.session()` was removed in https://github.com/pantsbuild/pants/pull/11685.